doc/source/user_guide/indexing.rst

```diff
@@ -1480,17 +1480,24 @@ default value.
    s.get('a')  # equivalent to s['a']
    s.get('x', default=-1)
 
-The :meth:`~pandas.DataFrame.lookup` method
--------------------------------------------
+.. _indexing.lookup
+
+Looking up values by index/column labels
+----------------------------------------
 
 Sometimes you want to extract a set of values given a sequence of row labels
-and column labels, and the ``lookup`` method allows for this and returns a
-NumPy array.  For instance:
+and column labels, this can be achieved by ``DataFrame.melt`` combined by filtering the corresponding
+rows with ``DataFrame.query``.  For instance:
 
 .. ipython:: python
 
-  dflookup = pd.DataFrame(np.random.rand(20, 4), columns = ['A', 'B', 'C', 'D'])
-  dflookup.lookup(list(range(0, 10, 2)), ['B', 'C', 'A', 'B', 'D'])
+    df = pd.DataFrame({'col': ["A", "A", "B", "B"],
+                       'A': [80, 23, np.nan, 22],
+                       'B': [80, 55, 76, 67]})
+    df
+    melt = df.melt('col')
+    df['lookup'] = melt.query('col == variable')['value'].to_numpy()
```

Comment:
rather than query, use .loc here as its more idiomatic (and plus its a single statement). Do we actually need to convert to numpy? (e.g. to avoid alignment)

Reply:
The reason I used `to_numpy` is indeed for the index alignment. With `.loc`  and without `to_numpy`, I would need `reset_index`. So that would look like:

```
melt = df.melt('col')
melt.loc[melt['col'] == melt['variable'], 'value'].reset_index(drop=True)
```

Reply:
Which one do you prefer? I can understand that this is more idiomatic, althought it's a bit more code.

Reply:
Btw now I'm typing it, I realize this method would fail if the reset was not `0, .., n-1`

Reply:
This is a horrible choice as an alternative for `lookup`. Essentially, `melt = df.melt('col')` duplicates the data (not to mention the extra `variable` column) while one can resolve to numpy indexing.

I'm  not sure what's the implementation of `lookup`, but it's a real request emerging from what I could see on SO. To me `pivot` and `pivot_table` are more redundant than `lookup`.

Reply:
@quangngd Let's not deprecate `pivot` and `pivot_table` now.

Reply:
> Which one do you prefer? I can understand that this is more idiomatic, althought it's a bit more code.

I do not think that this approach is correct. For example,  consider the following case

```
import pandas as pd
df = pd.DataFrame.from_dict({"V1": [66, 57, 79,75], "V2": [41,85,94,71], 
                             "V3":[19,3,38,58], "Selection":['V1','V3', 'V2','V3']})
 
melt = df.melt('Selection')
melt = melt.loc[melt['Selection'] == melt['variable'], 'value']
df['lookup'] = melt.reset_index(drop=True)
df
```

And you get a wrong output:
```

	V1	V2	V3	Selection	lookup
0	66	41	19	V1	66
1	57	85	3	V3	94
2	79	94	38	V2	3
3	75	71	58	V3	58

```

Where with the look up, you could get the correct answer with one command

Reply:
This is not a "implementation" @ quang. It it just one of the many approaches one can take to do a lookup in pandas, while making use of the already available methods. Maybe numpy indexing is a better choice (not sure how readable it is for newers users), so feel free to open a PR or ticket instead of commenting "horrible" on a closed PR. Pandas is open source, people spend their free time out of best interest trying to make it better, so can you.

Reply:
I never said it's an implementation. I said it's an **alternative**, and already explained why it's very bad. And it shows in the other (closed) [issue](https://github.com/pandas-dev/pandas/issues/39171#issuecomment-773374707) that it in fact is 1) more code, 2) error prone, 3) slower, and 4) less practical than even a vanilla `for` loop. I wouldn't have said that if it weren't be the suggested **alternative** on the official documentation. 

*Why wouldn't I go on and make a pull request?* From what I learned from the other discussion, you are not the one that has a say on bringing back `lookup`, and those that do don't seem to be willing to review the issue. Open source software thrives because not just people spend their free time writing it, but also people use and promote it. When the former stops listening to the needs of the latter and suggests lower-quality code based on [wrong assumption](https://github.com/pandas-dev/pandas/issues/39171#issuecomment-778936909), what's the point to contribute?

Reply:
@erfannariman and @quanghm 

i am not completely averse to bringing some form of lookup back

but it would have to be performant and not be yet another indexing api

if anyone wants to explore it great

@erfannariman is right here in that open source functions because the community not only what's something, but someone steps up to actually implement it

pandas core devs can provide review

Reply:
@jreback with all due respect, I'm surprised that you response to this thread after ignoring my questions in the other issue. I already [mentioned](https://github.com/pandas-dev/pandas/issues/39171#issuecomment-778936909) that `lookup` itself **is** an indexing API, equivalent to that of numpy. I don't really get your point when you say

> but it would have to be performant and **not be yet another indexing api**

Regarding the performance, the deprecated `lookup` implementation is as performant as it gets, sans some usual book-keeping when moving from numpy to pandas. And it's twice as fast as the **suggested alternative** across the various use cases. What exactly do you mean by *performant* here?

And please, 

>  but someone steps up to actually implement it

`lookup` **was** already implemented, and deprecated just because of what it does (indexing). Please correct me if I'm wrong (see arguments above), your criteria above just plainly say all efforts to implement `lookup` would be rejected.

All that said, if you decide not to bring back `lookup`, please replace the suggested `melt` method with just a single `for` loop. It'll be more readable to the new users while more *performant*. Not that I wanted any credit, but I **did** spend time to test and to discuss the issue, so I don't think it's fair to say 

> Pandas is open source, people spend their free time out of best interest trying to make it better, so can you.

in response to my post.

Reply:
@quanghm my response is to the tone of the argument

Reply:
i am not averse to the function of lookup but it's not going to be a top level api

Reply:
> i am not averse to the function of lookup but it's not going to be a top level api

Yet another reason why I don't go and make a pull request. I don't understand what's other than *top level api* that I can implement `lookup` functionality, and still don't understand *performant*.

> @quanghm my response is to the tone of the argument

Can you please be more specific? The way I see it, @erfannariman was basically saying that unless I made a pull request, I cannot criticize the suggested alternative as *horrible* even when I spent time to analyze, test and show that it is, in fact, horrible. OK, if *horrible* is an offensive word, my apology, English is not my first language. Let me re-phrase *it is very bad*. 

I do feel that my effort to contribute here is not valued. That was the best I can afford to better Pandas. Not everyone can afford to make a pull request, modify the code, run and pass all the unit tests.

Thanks for all the work maintaining and developing Pandas anyway.

Reply:
> From what I learned from the other discussion, you are not the one that has a say on bringing back `lookup`, and those that do don't seem to be willing to review the issue.

Opening a ticket with everything you wrote here has more chance to get reviewed by the core devs than to write all this on a closed PR. That way the other devs can see it as well, since I am quite sure they are not seeing this whole discussion.

If you don't feel comfortable open an issue, please let me know and I wil open one, since Jeff mentioned he's open for a (discussion) for re-implementing lookup.

Reply:
Actually I already opened the ticket since it's better to have the discussion there: see #40140 @quanghm @jreback